public/app/features/correlations/components/Wizard/wizardContext.tsx

```diff
@@ -0,0 +1,52 @@
+import React, { createContext, PropsWithChildren, useContext, useState } from 'react';
+import { FieldValues } from 'react-hook-form';
+
+export type WizardContextProps<T> = {
+  currentPage: number;
+  nextPage: () => void;
+  prevPage: () => void;
+  isLastPage: boolean;
+  onSubmit: (data: T) => void;
+  CurrentPageComponent: React.ComponentType;
+};
+
+export const WizardContext = createContext<WizardContextProps<FieldValues>>({
+  currentPage: 0,
+  nextPage: () => {},
+  prevPage: () => {},
+  isLastPage: true,
+  onSubmit: () => {},
+  CurrentPageComponent: () => null,
+});
+
+/**
+ * Dependencies provided to Wizard component required to build WizardContext
+ */
+type WizardContextProviderDeps<T> = {
+  pages: React.ComponentType[];
+  onSubmit: (data: T) => void;
+};
+
+/**
+ * Context providing current state and logic of a Wizard. Can be used by pages and navigation components.
+ */
+export function WizardContextProvider<T>(props: PropsWithChildren<WizardContextProviderDeps<T>>) {
+  const [currentPage, setCurrentPage] = useState(0);
+  const { pages, onSubmit, children } = props;
+
+  const context: WizardContextProps<T> = {
+    currentPage,
+    CurrentPageComponent: pages[currentPage],
+    isLastPage: currentPage === pages.length - 1,
+    nextPage: () => setCurrentPage(currentPage + 1),
+    prevPage: () => setCurrentPage(currentPage - 1),
+    onSubmit,
+  };
+
+  // @ts-ignore
```

Comment:
yeah, typing properly form contexts in this cases is quite weird. could we use `// @ts-expect-error` instead and a commend explaining why we are doing it?

Reply:
even better would be to not create a variable beforehand so we only suppress the check for `onSubmit`:


```tsx
return (
    <WizardContext.Provider
      value={{
        currentPage,
        CurrentPageComponent: pages[currentPage],
        isLastPage: currentPage === pages.length - 1,
        nextPage: () => setCurrentPage(currentPage + 1),
        prevPage: () => setCurrentPage(currentPage - 1),
       // @ts-expect-error ....
        onSubmit,
      }}
    >
      {children}
    </WizardContext.Provider>
  );
```

Reply:
even better would be to not create a variable beforehand so we only suppress the check for `onSubmit`:


```tsx
return (
    <WizardContext.Provider
      value={{
        currentPage,
        CurrentPageComponent: pages[currentPage],
        isLastPage: currentPage === pages.length - 1,
        nextPage: () => setCurrentPage(currentPage + 1),
        prevPage: () => setCurrentPage(currentPage - 1),
       // @ts-expect-error ....
        onSubmit,
      }}
    >
      {children}
    </WizardContext.Provider>
  );
```

Reply:
Makes way more sense. Changed in f1924d2cbcbeeca0d9681ae2cdf29f4b442d3c15